src/Products/PluginIndexes/CompositeIndex/CompositeIndex.py

```diff
@@ -0,0 +1,599 @@
+        """ wrapper to handle indexing of multiple attributes """
+
+        res = self._index_object(documentId, obj, threshold)
+        return res
+
+    def _index_object(self, documentId, obj, threshold=None):
+        """ index an object 'obj' with integer id 'i'
+
+        Ideally, we've been passed a sequence of some sort that we
+        can iterate over. If however, we haven't, we should do something
+        useful with the results. In the case of a string, this means
+        indexing the entire string as a keyword."""
+
+        # First we need to see if there's anything interesting to look at
+        # self.id is the name of the index, which is also the name of the
+        # attribute we're interested in.  If the attribute is callable,
+        # we'll do so.
+
+        # get permuted keywords
+        newKeywords = self._get_object_keywords(obj)
+
+        oldKeywords = self._unindex.get(documentId, None)
+
+        if oldKeywords is None:
+            # we've got a new document, let's not futz around.
+            try:
+                for kw in newKeywords:
+                    self.insertForwardIndexEntry(kw, documentId)
+                if newKeywords:
+                    self._unindex[documentId] = list(newKeywords)
+            except TypeError:
+                return 0
+        else:
+            # we have an existing entry for this document, and we need
+            # to figure out if any of the keywords have actually changed
+            if type(oldKeywords) is not OOSet:
+                oldKeywords = OOSet(oldKeywords)
+            newKeywords = OOSet(newKeywords)
+            fdiff = difference(oldKeywords, newKeywords)
+            rdiff = difference(newKeywords, oldKeywords)
+            if fdiff or rdiff:
+                # if we've got forward or reverse changes
+                self._unindex[documentId] = list(newKeywords)
+                if fdiff:
+                    self.unindex_objectKeywords(documentId, fdiff)
+                if rdiff:
+                    for kw in rdiff:
+                        self.insertForwardIndexEntry(kw, documentId)
+
+        return 1
+
+    def unindex_objectKeywords(self, documentId, keywords):
+        """ carefully unindex the object with integer id 'documentId'"""
+
+        if keywords is not None:
+            for kw in keywords:
+                self.removeForwardIndexEntry(kw, documentId)
+
+    def unindex_object(self, documentId):
+        """ carefully unindex the object with integer id 'documentId'"""
+
+        keywords = self._unindex.get(documentId, None)
+        self.unindex_objectKeywords(documentId, keywords)
+        try:
+            del self._unindex[documentId]
+        except KeyError:
+            LOG.debug('%s: Attempt to unindex nonexistent '
+                      'document with id %s' %
+                      (self.__class__.__name__, documentId),
+                      error=sys.exc_info())
+
```

Comment:
The above three functions looks almost identical to the KeywordIndex. Would it be possible to inherit from KeywordIndex instead of UnIndex? That might cut down on the code duplication of those methods.